File: modules/crypto/python/src/vfhe/crypto/prng.py
# SPDX-FileCopyrightText: 2026 Antonio Guimarães <antonio.guimaraes@imdea.org>
# SPDX-License-Identifier: Apache-2.0
"""The library's randomness, and the only randomness the library uses.

Two independent generators live in `c/src/prng.c` and are exposed here as two
objects, because the choice between them is a correctness question and not a
performance one:

- `entropy` (`EntropyPRNG`) — unpredictable bytes, seeded from RDRAND or
  `/dev/urandom` and expanded with the build's PRF. Use it for secrets:
  keys, noise, masks, anything an adversary must not guess.
- `seeded` (`SeededPRNG`) — values uniform in `[0, bound)` that are a pure
  function of `(context, seed)`. Use it where a value must be recomputable
  by another party from data it already has, such as a Fiat-Shamir
  transcript.

Both are module-level singletons over process-global C state; the classes
exist so the surface is typed and documented in one place, not so callers can
own an instance. Nothing here is thread-safe (the C pool is shared mutable
state).

`context` is a domain-separation tag: one seed under two different tags gives
two independent streams. Pass a fixed string literal per call site, never
anything caller-controlled.
"""

from __future__ import annotations

from contextlib import contextmanager
from typing import TYPE_CHECKING

from vfhe.engine import ffi, lib

if TYPE_CHECKING:
    from collections.abc import Iterator

# What `generate_rnd_seed` fills, in 64-bit words.
SEED_WORDS = 4

# The AES keystream's step, in bytes: the smallest amount `get_rnd_from_hash`
# is defined for.
_KEYSTREAM_STEP = 256


class EntropyPRNG:
    """The entropy-backed stream: bytes no one can predict.

    Every method draws from the process-global generator in `prng.c`, so two
    calls never return the same thing (short of the test-only override in
    `deterministic`).
    """

    def bytes(self, amount: int) -> bytes:
        """`amount` unpredictable bytes — the default entry point.

        Serves requests under 512 bytes from an internal pool and expands a
        fresh seed for larger ones.
        """
        if amount < 0:
            raise ValueError(f"amount must be non-negative, got {amount}")
        if amount == 0:
            return b""
        out = ffi.new("uint8_t[]", amount)
        lib.generate_random_bytes(amount, out)
        return bytes(out)

    def below(self, bound: int) -> int:
        """One value uniform in `[0, bound)`, from fresh entropy.

        Rejection sampling on whole bytes, so the result is unbiased. For a
        value that must be reproducible from a transcript, use `seeded`.
        """
        if bound < 1:
            raise ValueError(f"bound must be at least 1, got {bound}")
        if bound == 1:
            return 0
        width = (bound - 1).bit_length()
        nbytes = (width + 7) // 8
        mask = (1 << width) - 1
        while True:
            value = int.from_bytes(self.bytes(nbytes), "little") & mask
            if value < bound:
                return value

    def normal(self, sigma: float) -> float:
        """One sample from a zero-mean Gaussian with standard deviation
        `sigma` (Box-Muller over `bytes`).

        The support is unbounded: a caller needing a tail bound must clamp or
        resample.
        """
        return lib.generate_normal_random(sigma)

    def seed_words(self) -> list[int]:
        """One raw seed: `SEED_WORDS` words straight from the hardware source.

        The seed the expanders below draw from. Callers wanting random data
        want `bytes`, not this.
        """
        out = ffi.new("uint64_t[]", SEED_WORDS)
        lib.generate_rnd_seed(out)
        return [out[i] for i in range(SEED_WORDS)]

    def bytes_from_fresh_seed(self, amount: int) -> bytes:
        """`amount` bytes expanded from one freshly drawn seed, bypassing the
        pool. Every call pays for a seed draw, which dominates small requests.

        The AES keystream behind this on tuned x86-64 produces whole 256-byte
        steps, so a shorter request draws one step and keeps its prefix.
        """
        if amount < 1:
            raise ValueError(f"amount must be at least 1, got {amount}")
        width = max(amount, _KEYSTREAM_STEP)
        out = ffi.new("uint8_t[]", width)
        lib.get_rnd_from_hash(width, out)
        return bytes(out)[:amount]

    def bytes_from_pool(self, amount: int) -> bytes:
        """`amount` bytes from the 1 KiB pool, refilling it when what remains
        will not cover the request. A request larger than the pool is served
        from a fresh seed.
        """
        if amount < 1:
            raise ValueError(f"amount must be at least 1, got {amount}")
        out = ffi.new("uint8_t[]", amount)
        lib.get_rnd_from_buffer(amount, out)
        return bytes(out)

    @contextmanager
    def deterministic(self, seed: int) -> Iterator[None]:
        """Test-only: pin this stream to a splitmix64 sequence from `seed`.

        Makes every draw above reproducible for the duration of the block,
        within one build only — the expander is engine-dependent, so the same
        seed gives different bytes on different engines. Does not affect
        `seeded`, which is already a pure function of its arguments. Restores
        hardware entropy on exit.
        """
        lib.vfhe_prng_set_deterministic_seed(seed)
        try:
            yield
        finally:
            lib.vfhe_prng_clear_deterministic_seed()


class SeededPRNG:
    """The seeded sampler: values reproducible from `(context, seed)`.

    Byte-for-byte identical across runs and across engines, and untouched by
    `EntropyPRNG.deterministic`. This is what a protocol uses for values both
    parties must agree on.
    """

    def below(self, bound: int, context: bytes, seed: bytes) -> int:
        """One value uniform in `[0, bound)`."""
        return self.below_many(1, bound, context, seed)[0]

    def below_many(
        self, count: int, bound: int, context: bytes, seed: bytes, start: int = 0
    ) -> list[int]:
        """`count` values uniform in `[0, bound)`, from position `start` of
        the sequence.

        Entry `i` depends on `i` alone, so raising `count` extends the
        sequence instead of changing it, and a window costs what its own
        values cost: ``below_many(k, ..., start=j)`` is
        ``below_many(j + k, ...)[j:]`` without producing the first `j`.
        """
        if bound < 1:
            raise ValueError(f"bound must be at least 1, got {bound}")
        if count < 0:
            raise ValueError(f"count must be non-negative, got {count}")
        if start < 0:
            raise ValueError(f"start must be non-negative, got {start}")
        if count == 0:
            return []
        out = ffi.new("uint64_t[]", count)
        lib.prng_sample_below_from(out, count, start, bound, context, seed, len(seed))
        return [out[i] for i in range(count)]

    def bytes(self, amount: int, context: bytes, seed: bytes) -> bytes:
        """`amount` bytes, a pure function of `(context, seed)`.

        Drawn as `amount` uniform bytes from `below_many`, so this is the
        seeded counterpart of `EntropyPRNG.bytes` and not a second expander.
        """
        if amount < 0:
            raise ValueError(f"amount must be non-negative, got {amount}")
        if amount == 0:
            return b""
        return bytes(self.below_many(amount, 256, context, seed))


entropy = EntropyPRNG()
seeded = SeededPRNG()

File: modules/misc/c/src/prng.c
// SPDX-FileCopyrightText: 2026 Antonio Guimarães <antonio.guimaraes@imdea.org>
// SPDX-License-Identifier: Apache-2.0
#include "misc.h"
#if defined(__x86_64__) || defined(_M_X64) || defined(__i386__) || defined(_M_IX86)
#include <immintrin.h>
#endif
#include <assert.h>
#include <math.h>
#include <stdbool.h>
#include <stdint.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#include <blake3.h>

#ifdef USE_SHAKE
#include "../third-party/sha3/fips202.c"
#endif

// Forward declaration of aes_prng (only used if AES-NI is available and not in portable build)
#if !defined(PORTABLE_BUILD) && !defined(PORTABLE) && (defined(__x86_64__) || defined(_M_X64)) &&  \
    defined(__AES__)
void aes_prng(uint8_t *output, uint64_t outlen, const uint8_t *input, uint64_t inlen);
#endif

// --- Deterministic seed override (tests only) ----------------------------
// Reproducible FHE-bootstrap tests need a fixed RNG. When a deterministic seed
// is set (vfhe_prng_set_deterministic_seed), generate_rnd_seed yields a
// reproducible, non-repeating stream via splitmix64 instead of hardware
// entropy. Production never calls the setters, so it keeps using RDRAND /
// /dev/urandom. Not thread-safe; intended for single-threaded test use.
static int det_active = 0;
static uint64_t det_state = 0;

// Buffered entropy pool (lifted to file scope so the setters can discard it).
static uint8_t rnd_buffer[1024] __attribute__((aligned(64)));
static uint64_t rnd_buffer_idx = sizeof(rnd_buffer);

static uint64_t splitmix64_next(uint64_t *s)
{
    uint64_t z = (*s += 0x9E3779B97F4A7C15ULL);
    z = (z ^ (z >> 30)) * 0xBF58476D1CE4E5B9ULL;
    z = (z ^ (z >> 27)) * 0x94D049BB133111EBULL;
    return z ^ (z >> 31);
}

static void det_fill_seed(uint64_t *p)
{
    for (int i = 0; i < 4; i++)
        p[i] = splitmix64_next(&det_state);
}

#if (defined(__x86_64__) || defined(_M_X64)) && !defined(PORTABLE) && !defined(PORTABLE_BUILD)
// Intel hardware RDRAND seed generator
void generate_rnd_seed(uint64_t *p)
{
    if (det_active)
    {
        det_fill_seed(p);
        return;
    }
    if (0 == _rdrand64_step((unsigned long long *)p) ||
        0 == _rdrand64_step((unsigned long long *)&(p[1])) ||
        0 == _rdrand64_step((unsigned long long *)&(p[2])) ||
        0 == _rdrand64_step((unsigned long long *)&(p[3])))
    {
        printf("Random Generation Failed\n");
        return;
    }
}
#else
// Fallback urandom seed generator
void generate_rnd_seed(uint64_t *p)
{
    if (det_active)
    {
        det_fill_seed(p);
        return;
    }
    FILE *fp = fopen("/dev/urandom", "r");
    const size_t read = fp ? fread(p, 1, 32, fp) : 0;
    if (fp)
        fclose(fp);
    if (read != 32)
    {
        printf("Random Generation Failed\n");
        return;
    }
}
#endif

void get_rnd_from_hash(uint64_t amount, uint8_t *pointer)
{
    uint64_t rnd[4];
    generate_rnd_seed(rnd);

#if !defined(PORTABLE_BUILD) && !defined(PORTABLE) && (defined(__x86_64__) || defined(_M_X64)) &&  \
    defined(__AES__)
    aes_prng(pointer, amount, (uint8_t *)rnd, 32);
#elif defined(USE_SHAKE)
    shake256(pointer, amount, (uint8_t *)rnd, 32);
#else
    // Default fallback is Blake3
    blake3_hasher hasher;
    blake3_hasher_init(&hasher);
    blake3_hasher_update(&hasher, (uint8_t *)rnd, 32);
    blake3_hasher_finalize(&hasher, pointer, amount);
#endif
}

void get_rnd_from_buffer(uint64_t amount, uint8_t *pointer)
{
    // A request the pool could never satisfy is served directly, so the copy
    // below always stays inside it.
    if (amount > sizeof(rnd_buffer))
    {
        get_rnd_from_hash(amount, pointer);
        return;
    }
    if (amount > (sizeof(rnd_buffer) - rnd_buffer_idx))
    {
        rnd_buffer_idx = 0;
        get_rnd_from_hash(sizeof(rnd_buffer), rnd_buffer);
    }
    memcpy(pointer, rnd_buffer + rnd_buffer_idx, (size_t)amount);
    rnd_buffer_idx += amount;
}

void generate_random_bytes(uint64_t amount, uint8_t *pointer)
{
    if (amount < 512)
        get_rnd_from_buffer(amount, pointer);
    else
        get_rnd_from_hash(amount, pointer);
}

// Test-only: pin the PRNG to a reproducible stream (see the note above). The
// setter also discards any buffered entropy so the next draw starts from the
// seed; clearing returns to hardware entropy.
void vfhe_prng_set_deterministic_seed(uint64_t seed)
{
    det_active = 1;
    det_state = seed;
    rnd_buffer_idx = sizeof(rnd_buffer);
}

void vfhe_prng_clear_deterministic_seed(void)
{
    det_active = 0;
    rnd_buffer_idx = sizeof(rnd_buffer);
}

// --- Gaussian sampling ---------------------------------------------------
// Box-Muller over the entropy-backed stream above, so it follows the
// deterministic-seed override too. Moved here from misc_tp.c: every generator
// in the library lives in this file.

double int2double(uint64_t x) { return ((double)x) / 18446744073709551616.0; }

#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

double generate_normal_random(double sigma)
{
    uint64_t rnd[2];
    generate_random_bytes(16, (uint8_t *)rnd);
    return cos(2. * M_PI * int2double(rnd[0])) * sqrt(-2. * log(int2double(rnd[1]))) * sigma;
}

// --- Seeded sampling (deterministic, independent of the stream above) ----
// A *different* generator from everything above: the caller supplies the seed,
// so the result is a pure function of (context, seed) and the
// deterministic-seed override has nothing to do with it. Used where a value
// must be reproducible from a transcript rather than merely unpredictable.
//
// `context` is a domain-separation tag: two callers with the same seed and
// different tags get independent streams. Pass a fixed string literal per use
// site, never anything caller-controlled.

void prng_sample_below(uint64_t *out, uint64_t count, uint64_t bound, const char *context,
                       const uint8_t *seed, uint64_t seed_len)
{
    assert(bound > 0); // otherwise the rejection loop below never terminates

    blake3_hasher hasher;
    blake3_hasher_init_derive_key(&hasher, context);
    blake3_hasher_update(&hasher, seed, seed_len);

    // Smallest 2^k - 1 that covers `bound`, so rejection discards under half
    // the draws.
    uint64_t mask = bound;
    mask |= mask >> 1;
    mask |= mask >> 2;
    mask |= mask >> 4;
    mask |= mask >> 8;
    mask |= mask >> 16;
    mask |= mask >> 32;

    // BLAKE3 as an XOF: `blake3_hasher_finalize` is a pure function of hasher
    // state, so finalizing repeatedly without an intervening update returns the
    // same bytes. Seeking along the output stream is what makes successive
    // draws independent.
    uint64_t offset = 0;
    for (uint64_t i = 0; i < count; i++)
    {
        for (;;)
        {
            uint8_t buf[sizeof(uint64_t)];
            blake3_hasher_finalize_seek(&hasher, offset, buf, sizeof(buf));
            offset += sizeof(buf);

            uint64_t word;
            memcpy(&word, buf, sizeof(word)); // the buffer is not aligned for a cast
            const uint64_t sampled = word & mask;
            if (sampled < bound)
            {
                out[i] = sampled;
                break;
            }
        }
    }
}
